teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
+  (no flags, just flag 1, flags 1 & 2, flags 1 & 3, just flag 2, flags 3 & 3,
+  just flag 3, all flags)
+
+### Maintain multiple verisons of impacted CRDs
+
+In this version, if we want to add an alpha field to a beta type, we would create
+a new alpha type and we would serve both.
+
+For example, right now Task has `v1alpha1` (deprecated) and `v1beta1`. We could also
+create a version such as `v1alpha2` and add new fields to that version before promoting them
+to `v1beta1`.
+
+Pros:
+* From the user's point of view, this might be the most intuitive option:
+  if they want to opt into alpha features for a beta CRD (e.g. Task is currently v1beta1),
+  they use the alpha version (v1alpha1)
+* Operators can control access by not installing CRD versions they don't want being used
+
+Cons:
+* The maintenance burden for Tekton contributors is the biggest blocker: every time we add
+  a new CRD version we need to do a lot of copy and paste of both code and tests
+      * Potentially we could mitigate this by creating tooling to make it easier?
+* Will need another approach for changes in behavior
+* There is still only one storage type; if we wanted to avoid changes to the storage type
+  we'd have to leverage something like annotations to store the new fields
+* Not clear if we'd need to make a new version every time we add a new field; or have one version
+  with all experimental fields
+
+### Documentation only
+
+In this version, we make no changes to functionality, but we make very clear documentation
+that shows which fields are considered beta.
+
+Pros:
+* No code changes
+
+Cons:
+* If people don't look at the docs, they can easily be surprised by changes to "alpha" fields
+  * These will still look like backwards incompatible changes
+* Doesn't help with changes in behavior
+
+### Warn on use of new fields
+
+Handling this with warnings (e.g. warn the user in logs when they use alpha features)
+
+Pros:
+* Few code changes
+
+Cons:
+* If people don't look at the logs (and when do they do that!, they can easily be
+  surprised by changes to "alpha" fields
+  * These will still look like backwards incompatible changes
+  * One way to mitigate this would be to create a way to communicate warnings like this within
+    the status of objects
+* Doesn't help with changes in behavior
+
+### Indicate stability level with field name
+
+In this option we create a naming convention that allows us to indicate the stability level with
+the name and rename them when the fields graduate.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: task-with-parameters
+spec:
+  steps:
+    - name: this-can-fail
+      image: some-image
+      args: ["some-arg"]
+      alpha_allow_failure: true # starting the field name with alpha indicates it has alpha stablity
+```
+
+Pros:
+* Very clear when a field is alpha and that changes will be required once it is stable
```

Comment:
Additional pro: I think the Conversion Webhook story for fields like this is very clear as they migrate up to and down from stored version: alpha->beta conversion upgrades `alpha_*` to `beta_*` field names and beta->v1 conversion updates `beta_*` to just `*`.